MRG: Powerlaw Kolmogorov–Smirnov test

hduongtrong opened the pull request:
1. Perform the Kolmogorov-Smirnov test on the distances travelled in 20ms by a mouse of a strain in a day
2. Create subfolder for the Powerlaw subproject.

This commit is similar to Pull Request #149 , except this one uses Kolmogorov-Smirnov test instead of Generalized Likelihood Ratio test.

hduongtrong commented:
@doutib @FengshiNiu @CenzhuoYao @lunadadada @z357412526 was wondering if you guys can take a look and leave some comments for possible places to improve?
I create a subfolder for our subproject as well.

doutib commented on mousestyles/powerlaw/kolmogorov_test.py:
```diff
@@ -0,0 +1,160 @@
+
+    # Fit the Parameter by explicitely minimize KS-Statistics, and perform
+    # K-S test.
+    def calculate_ks_statistics(params):
```
For clarity purposes, I would define this function outside of the perform_kstest function.

doutib commented:
- I would define calculate_ks_statistics outside of the main function.
- I would delete the plotting option and make another script in the plots folde for example.
- why [4.5, 0.5, 0.5] is a good initialization for the pareto, isn't is better to give this as a default parameter?
- I would define many test functions instead of only one because it is far easier to debug in case of problems.

CenzhuoYao commented on mousestyles/powerlaw/kolmogorov_test.py:
```diff
@@ -0,0 +1,160 @@
+
+    if dist_name == "pareto":
+        # A good initial estimate for Pareto parameter
+        init_params = [4.5, 0.5, 0.5]
```
Could you explain why [4.5, 0.5, 0.5] is a good initial estimate for Pareto distribution. Also, if not Pareto, what is distribution.fit function? Is that defined before?

hduongtrong replied:
Sure. so distribution is a scipy.stats.rv_continuous object, and the distribution.fit will fit maximum likelihood parameter to the data. See for example scipy doc [http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.rv_continuous.fit.html](url)

hduongtrong commented:
@doutib @CenzhuoYao thanks for reviewing the codes. 
1. About the initial [4.5, 0.5, 0.5] initial estimate, what I'm doing is fit the parameter to Pareto distribution by finding the Pareto CDF that is closest to our sample CDF. I can define a function which takes in parameter for Pareto, and return the distance between this Pareto distribution and our sample distribution. I minimize this function using scipy.optimize, and for this I need an initial guess. It we just put in bad initial guess, the function won't be able to converge to a good optimal value (i.e. the found CDF is not close to the sample CDF). I found that this initial parameter will help scipy.optimize find a good optimal value.

hduongtrong commented:
To add to earlier, so the point is if my optimizer is good enough, it will be able to find the optimal parameter, doesn't matter what my initial estimate is. But here sometimes it get stuck in a local optima, so I help it by giving it a good starting point to search for good parameter.

2. About moving plot function: good point, I've moved it to the doc/source/report dicrectory
3. More test function: good point, I've added a couple more tests. The test coverage is 100%

hduongtrong commented:
I've made some changes as suggested. If you guys can take a quick look that'd be great, so I can change this to MPR status.

doutib commented on doc/source/report/plots/hist_kolmogorov.py:
```diff
@@ -0,0 +1,42 @@
+    mle_data = distribution(*mle_params).rvs(len(distances))
+    opt_data = distribution(*opt_params).rvs(len(distances))
+
+    bins = np.linspace(min(distances), max(distances), 50)
```
I would define 50 as a default parameter instead.

doutib commented on mousestyles/distribution/kolmogorov_test.py:
```diff
@@ -0,0 +1,147 @@
+    if dist_name == "pareto":
+        # A good initial estimate for Pareto parameter to help optimizer
+        # converge to a better parameter
+        init_params = [4.5, 0.5, 0.5]
```
Can you add the init_params as a default parameter too here?

hduongtrong replied:
there is a problem because this init_params is only good for Pareto distribution. If I have that as a default parameter, and user uses a different distribution (e.g. exponential), then the result will be bad. For other distribution, initialize the search with the mle parameters is better.

doutib commented on mousestyles/distribution/test/test_kolmogorov_test.py:
```diff
@@ -0,0 +1,27 @@
+    assert type(res) is np.ndarray
+    assert res.shape == (3,)
+    assert np.allclose(res, np.array([3.67593246, 0.62795748, 0.37224205]))
+
```
You can split this function into two I guess.

hduongtrong replied:
that makes sense

doutib commented on mousestyles/distribution/test/test_kolmogorov_test.py:
```diff
@@ -0,0 +1,27 @@
+                         distribution=stats.pareto, verbose=False)
+    assert type(res) is np.ndarray
+    assert res.shape == (3,)
+    assert np.allclose(res, np.array([3.67593246, 0.62795748, 0.37224205]))
```
I don't think this is a good test to include because the values [3.67593246, 0.62795748, 0.37224205] don't really mean anything a priori. You should just remove this.

doutib commented on mousestyles/distribution/test/test_kolmogorov_test.py:
```diff
@@ -0,0 +1,27 @@
+def test_get_travel_distances():
+    res = get_travel_distances(strain=0, mouse=0, day=0)
+    assert type(res) is np.ndarray
+    assert len(res) == 38845
```
Does this length mean anything? If not I think it is better not to include it. You should instead make tests based on prior facts that you know to be true.

hduongtrong commented:
thanks a lot @doutib for the comments. I'll go over and fix them.

hduongtrong commented:
thanks again @doutib and @CenzhuoYao for the comments. I've fixed all of them, except the part of moving init_params to default variable. I tried different way to get rid of it, e.g. I can first do a brute force optimization to find a init_params, then use this init_params to feed in a smarter optimization function to find the optimal. The problem is it will be pretty slow. Let me know if you have some suggestions.

If anyone has time to take a look at the updated codes and review it, that'd be great.

hduongtrong commented:
@shamindras @peterasujan @dylandaniels  my team has already helped me reviewing the codes, was wondering if you guys can help take a look and tell me possible places for improvement?

shamindras commented:
@hduongtrong - nice to see that you have done a code review.

Some general ideas would be to:

- Revisit the code and make sure that [it is pythonic](http://docs.python-guide.org/en/latest/writing/style/)
- See whether all tests are correctly documented with comments. Ask the basic question - if I were to look at this code in 6 days, 6 weeks and 6 months would I understand what it is doing? Better documentation would help ensure this, if it is not the case
- You can have a look at our [code review process](https://github.com/berkeley-stat222/mousestyles/pull/136/files) for some ideas. If you have any changes on our code please feel free to add them 😄 !

hduongtrong commented:
thanks a lot @shamindras for the suggestions. I've added even more comments and documentation as suggested. I'll change the status to MPR unless anybody else still have some ideas for improvements.

jarrodmillman commented on doc/source/report/plots/hist_kolmogorov.py:
```diff
@@ -0,0 +1,65 @@
+    dist_name = distribution.__class__.__name__.split("_")[0]
+    distances = get_travel_distances(strain, mouse, day)
+    mle_params = distribution.fit(distances)
+    opt_params = perform_kstest(strain, mouse, day, verbose=False)
```
This is syntactically invalid.  Your function takes 3 arguments and you provide 4.  Also please include this plot in the report.

jarrodmillman commented:
@hduongtrong You need to fix the figure for the report (see inline comments).  I've also merged other pull requests, so please rebase.  Once you've done that (changes this back to MRG and) I will check everything again, but I suspect I will merge it once you've done these two things.

hduongtrong commented:
thanks @jarrodmillman for the suggestions. I'll fix that and make another pull request.

jarrodmillman commented:
@hduongtrong Please add a plot directive to the report so I can check whether plotting script is correct.  As I mentioned in my previous comment, you _need_ to do this before I will merge your pr. Once you've done that rebase on upstream master and change this pr to MRG.

hduongtrong commented:
@jarrodmillman Sorry about that I misunderstood. I've added this line 
`.. plot:: report/plots/hist_kolmogorov.py`
to distribution.rst (together with caption). I assume that is sufficient, though I don't see any change to the report when seen on github. I also rebased.
